### PR TITLE
Scoping document for derived variables

### PR DESCRIPTION
**Pull request by yulric**
You will need to install some dependencies using `renv::restore` to run this document. The command to run it is, `quarto render ./docs/derived-variables.qmd`.

### CODE REVIEW

File: docs/derived-variables.qmd [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+### Function Language
+
+The library should support only functions written in R. The R ecosystem is moving 
+towards allowing its users to execute code from other statistical languages. An 
+example is the [reticulate](https://rstudio.github.io/reticulate/index.html)
+package which enables the execution of arbitrary 
+Python code within a .R file. Example below,
+
+```{r, python-in-R}
+# Python code that creates a function to calculate BMI
```
Comment (DougManuel) [inline docs/derived-variables.qmd:35]: I suggest the qmd should be executable- meaning add the reticulate and knitr library to this file.

**Review by DougManuel — DISMISSED** [review]
1. I'd probably summarize the scope at best practices early, but the info is in the document:
    1. Derived function should work with or without rec_with_table and the recodeflow package. Meaning, a user should be able to use a derived function in other applications, without need for rec_with_table. ("cut-and-paste" to other code).
    1. Derived functions in recodeflow handle scalar, vector, and data frame inputs.
        - Functions should prioritize vectorization for performance.
        - This means derived functions should execute on vectors.
    1. They require mathematical operations and conditional logic.
    1. Tagged NA values must be handled properly.
    1. Logging is supported, including simple logging. ** for discussion **
    1. Proper documentation, error handling, and testing are required.
        - max, min and similar attributes for error handling should not be hard coded. Rather, these attributes should be stored in variables and variable_details. 
    1. Only R-based functions are supported for now, but with provisions for additional languages.

1. Do we want to specify how a user would create and use derived functions for scalar, vector, and data frames? This may require discussion. It seems to me that we just say that function should work on vectors. (Sorry for polluting your code: I've added a few more methods for testing.)
    - I suggest that we are clear on the supported or preferred dependancies. I suggest for example that we are explicit in how tagged_NA() are supported. (But this is TBA for discussion). 
    - On the balance, I suggest we start using and supporting dpylr for examples, i.e. 
    
   ```
     test_data %>% 
    mutate(bmi = vector_BMI(weight, height))
    ```
    
    In the rec_with_table(), we'd require derived functions that work on vectors, but we'd probably want to implement the more efficient approach of
    ```
     vector_BMI(test_data$weight, test_data$height)
    ```
1. We could be a bit more explicit about best practices for documentation.
```Proper documentation is provided for each function and inlcude
	- Function behavior.
	- Expected inputs and outputs.
	- Example usage with scalar, vector, data frame inputs, and use with rec_with_table.
	- Test cases for edge cases (e.g., missing values, extreme values).

 1. We'll want to update these specifications for logging, once we have that discussion finished.

**Comment by DougManuel** [comment]
Sorry, I added performance examples, but the whole document was reformatted. We may want to decide on a linter and style approach and then save those to the repo. We could also add git actions if we want.

**Comment by DougManuel** [comment]
Here are the performance measure on my computer, but they seem to vary if I rerun them. Scalar is clearly not great, but all other approaches seem reasonable to me. "Vector" often performes much better and it seems reasonable to use that appraoch in rec_with_table. I think we should use dplyr in our vignettes and function examples if we want to show DF examples that don't use rec_with_table.

|Type                    | Run.Time.in.ms|Percent.Difference |
|:-----------------------|--------------:|:------------------|
|Scalar                  |     30.1218033|Reference          |
|Vector                  |      0.7650852|190.091779944268   |
|Vectorized              |      8.3370209|113.288863540556   |
|Mapply (vectorized_BMI) |     30.1349163|0.0435238490909451 |
|Mapply (vector_BMI)     |      3.6549568|156.716312557352   |
|Mapply (scalar_BMI)     |      0.8940697|188.469521100776   |
|Dplyr (vector_BMI)      |      0.7610321|190.142976252972   |
|Dplyr (vectorized_BMI)  |      1.4338493|181.824500959548   |

File: docs/derived-variables.qmd (hunk 1 above)
Comment (yulric) [reply docs/derived-variables.qmd:53]: I added them to the Suggests field in the DESCRIPTION file. Do you also mean to load them in using `library`?

**Comment by yulric** [comment]
Comments:

* `Only R-based functions are supported for now, but with provisions for additional languages.`. This was addressed in the function language section right?
* `I suggest that we are clear on the supported or preferred dependancies. I suggest for example that we are explicit in how tagged_NA() are supported`. I was going to leave that to the tagged NA scope document.
* `In the rec_with_table(), we'd require derived functions that work on vectors, but we'd probably want to implement the more efficient approach of`. Sorry what do you mean? If we're deciding that derived functions should work on vectors then rec_with_table should not need to do anything right?
* `We could be a bit more explicit about best practices for documentation.`. Could you expand on that and I'll add them in.

To discuss:

1. Whether functions  should support all input types (scalar, vector, and data.frame) or only one
2. Tagged NA values must be handled properly.
3. Logging is supported, including simple logging.

File: docs/derived-variables.qmd [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+vector_run_time <- get_func_perf(function() {
+  vector_BMI(test_data$weight, test_data$height)
+})
+
+vectorized_run_time <- get_func_perf(function() {
+  vectorized_BMI(test_data$weight, test_data$height)
+})
+
+# Mapply methods - testing all three functions
+mapply_vectorized_run_time <- get_func_perf(function() {
```
Comment (yulric) [inline docs/derived-variables.qmd:144]: @DougManuel Does it makes sense to run the mapply tests with the vector BMI functions? It will essentially run them as a scalar function passing in each value in the vector one at a time, defeating the purpose of those functions accepting a vector.

**Comment by yulric** [comment]
@DougManuel Can you tell me what program you're using to format the file? I'll start using it.

**Comment by DougManuel** [comment]
> @DougManuel Can you tell me what program you're using to format the file? I'll start using it.

I suggest we use the standard Tidyverse style guide. We can modify, but we should add configurations for any modifications. 

Tidyverse style guide calls out styler and lintR in their documentation, and I use those packages. https://style.tidyverse.org

There are GH Actions for both. https://github.com/r-lib/actions/tree/v2/examples See `lint.yaml` and `style.yaml`.  These have been added to chmsflow (under development). 

I do use other stylers in VSCode and Positron, and I sometimes worry they overwrite styleR, but that's my problem. I use Rewrap for VSCode, but it isn't supported in Positron.

**Comment by yulric** [comment]
Waiting on https://github.com/Big-Life-Lab/recodeflow/pull/50 to be merged in

**Comment by DougManuel** [comment]
The PR can be marked as completed, right?

**Review by DougManuel — APPROVED** [review]

**Comment by yulric** [comment]
> The PR can be marked as completed, right?

Yes its ready to be merged but I'm waiting for the upstream PRs to be merged first
